markdown_test.go

```diff
@@ -0,0 +1,135 @@
+
+// TestWriteError ensures that the cmd.PrintMarkdown function surfaces errors
+// returned by the writer.
+func (*markdownSuite) TestWriteError(c *gc.C) {
+	expectedErr := errors.New("foo")
+	writer := errorWriter{err: expectedErr}
+	command := &docTestCommand{
+		info: &cmd.Info{},
+	}
+	err := cmd.PrintMarkdown(writer, command, cmd.MarkdownOptions{})
+	c.Assert(err, gc.NotNil)
+	c.Check(err, gc.ErrorMatches, ".*foo")
+}
+
+// errorWriter is an io.Writer that returns an error whenever the Write method
+// is called.
+type errorWriter struct {
+	err error
+}
+
+func (e errorWriter) Write([]byte) (n int, err error) {
+	return 0, e.err
+}
+
+// TestOutput tests that the output of the PrintMarkdown function is
+// fundamentally correct.
+func (*markdownSuite) TestOutput(c *gc.C) {
+	seeAlso := []string{"clouds", "update-cloud", "remove-cloud", "update-credential"}
+	subcommands := map[string]string{
+		"foo": "foo the bar baz",
+		"bar": "bar the baz foo",
+		"baz": "baz the foo bar",
+	}
+
+	command := &docTestCommand{
+		info: &cmd.Info{
+			Name:        "add-cloud",
+			Args:        "<cloud name> [<cloud definition file>]",
+			Purpose:     "Add a cloud definition to Juju.",
+			Doc:         "details for add-cloud...",
+			Examples:    "examples for add-cloud...",
+			SeeAlso:     seeAlso,
+			Aliases:     []string{"new-cloud", "cloud-add"},
+			Subcommands: subcommands,
+		},
+		flags: []testFlag{{
+			name: "force",
+		}, {
+			name:  "file",
+			short: "f",
+		}, {
+			name:  "credential",
+			short: "c",
+		}},
+	}
+
+	// These functions verify the provided argument is in the expected set.
+	linkForCommand := func(s string) string {
+		for _, cmd := range seeAlso {
+			if cmd == s {
+				return "https://docs.com/" + cmd
+			}
+		}
+		c.Fatalf("linkForCommand called with unexpected command %q", s)
+		return ""
+	}
+
+	linkForSubcommand := func(s string) string {
+		_, ok := subcommands[s]
+		if !ok {
+			c.Fatalf("linkForSubcommand called with unexpected subcommand %q", s)
+		}
+		return "https://docs.com/add-cloud/" + s
+	}
+
+	var buf bytes.Buffer
+	err := cmd.PrintMarkdown(&buf, command, cmd.MarkdownOptions{
+		Title:             `Command "juju add-cloud"`,
+		UsagePrefix:       "juju ",
+		LinkForCommand:    linkForCommand,
+		LinkForSubcommand: linkForSubcommand,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	fmt.Println(buf.String())
+	c.Check(buf.String(), gc.Equals, `# Command "juju add-cloud"
+
+> See also: [clouds](https://docs.com/clouds), [update-cloud](https://docs.com/update-cloud), [remove-cloud](https://docs.com/remove-cloud), [update-credential](https://docs.com/update-credential)
+
+**Aliases:** new-cloud, cloud-add
+
+## Summary
+Add a cloud definition to Juju.
+
+## Usage
+`+"```juju add-cloud [options] <cloud name> [<cloud definition file>]```"+`
+
+### Options
+| Flag | Default | Usage |
+| --- | --- | --- |
+| `+"`-c`, `--credential`"+` | default value for "credential" flag | description for "credential" flag |
```

Comment:
Look at use test fixtures in a testdata directory to compare output to so we don't need this.

https://dave.cheney.net/2016/05/10/test-fixtures-in-go

Reply:
Good idea, done